src/services/refactors/convertToNamedParameters.ts

```diff
@@ -0,0 +1,519 @@
+        const sourceFile = sourceNode.getSourceFile();
+        const text = sourceFile.text;
+        if (hasLeadingLineBreak(sourceNode, text)) {
+            copyLeadingComments(sourceNode, targetNode, sourceFile);
+        }
+        else {
+            copyTrailingAsLeadingComments(sourceNode, targetNode, sourceFile);
+        }
+        copyTrailingComments(sourceNode, targetNode, sourceFile);
+
+        function hasLeadingLineBreak(node: Node, text: string) {
+            const start = node.getFullStart();
+            const end = node.getStart();
+            for (let i = start; i < end; i++) {
+                if (text.charCodeAt(i) === CharacterCodes.lineFeed) return true;
+            }
+            return false;
+        }
+    }
+
+    function getParameterName(paramDeclaration: ValidParameterDeclaration) {
+        return getTextOfIdentifierOrLiteral(paramDeclaration.name);
+    }
+
+    function getClassNames(constructorDeclaration: ValidConstructor): Identifier[] {
+        switch (constructorDeclaration.parent.kind) {
+            case SyntaxKind.ClassDeclaration:
+                const classDeclaration = constructorDeclaration.parent;
+                return [classDeclaration.name];
+            case SyntaxKind.ClassExpression:
+                const classExpression = constructorDeclaration.parent;
+                const variableDeclaration = constructorDeclaration.parent.parent;
+                const className = classExpression.name;
+                if (className) return [className, variableDeclaration.name];
+                return [variableDeclaration.name];
+        }
+    }
+
+    function getFunctionNames(functionDeclaration: ValidFunctionDeclaration): Node[] {
+        switch (functionDeclaration.kind) {
+            case SyntaxKind.FunctionDeclaration:
+            case SyntaxKind.MethodDeclaration:
+                return [functionDeclaration.name];
+            case SyntaxKind.Constructor:
+                const ctrKeyword = findChildOfKind(functionDeclaration, SyntaxKind.ConstructorKeyword, functionDeclaration.getSourceFile())!;
+                if (functionDeclaration.parent.kind === SyntaxKind.ClassExpression) {
+                    const variableDeclaration = functionDeclaration.parent.parent;
+                    return [variableDeclaration.name, ctrKeyword];
+                }
+                return [ctrKeyword];
+            case SyntaxKind.ArrowFunction:
+                return [functionDeclaration.parent.name];
+            case SyntaxKind.FunctionExpression:
+                if (functionDeclaration.name) return [functionDeclaration.name, functionDeclaration.parent.name];
+                return [functionDeclaration.parent.name];
+            default:
+                return Debug.assertNever(functionDeclaration);
+        }
+    }
+
+    type ValidParameterNodeArray = NodeArray<ValidParameterDeclaration>;
+
+    type ValidVariableDeclaration = VariableDeclaration & { name: Identifier, type: undefined };

```

Comment:
Consider making this a subtype of `VariableDeclaration`.  That would be more consistent with the other types and would result in better language service support (right now, the type is something like `Identifier | (BindingElement & Identifier)`).